lib/datadog/statsd/sender.rb

```diff
@@ -1,159 +1,130 @@
-            # a call from another thread has already re-created
-            # the companion thread before this one acquired the lock
+          @mux.synchronize {
+            return if @is_closed
+            # test if a call from another thread has already re-created
+            # the background thread before this one acquired the lock
             break if sender_thread.alive?
-            @logger.debug { "Statsd: companion thread is dead, re-creating one" } if @logger
 
-            message_queue.close if CLOSEABLE_QUEUES
-            @message_queue = nil
-            message_buffer.reset
-            start
+            # re-create the channel of communication since we will spawn a new bg thread
+            channel.close if CLOSEABLE_QUEUES
+            @channel = Queue.new
+            message_buffer.reset # don't use messages appended by another fork
+            @sender_thread = Thread.new(&method(:send_loop))
           }
         end
 
-        message_queue << message
+        channel << message
       end
 
-      def start
-        raise ArgumentError, 'Sender already started' if message_queue
-
-        # initialize a new message queue for the background thread
-        @message_queue = Queue.new
-        # start background thread
-        @sender_thread = Thread.new(&method(:send_loop))
+      # Compatibility with `Sender`
+      def start()
       end
 
-      if CLOSEABLE_QUEUES
-        # when calling stop, make sure that no other threads is trying
-        # to close the sender nor trying to continue to `#add` more message
-        # into the sender.
-        def stop(join_worker: true)
-          message_queue = @message_queue
-          message_queue.close if message_queue
-
-          sender_thread = @sender_thread
-          sender_thread.join if sender_thread && join_worker
-        end
-      else
-        # when calling stop, make sure that no other threads is trying
-        # to close the sender nor trying to continue to `#add` more message
-        # into the sender.
-        def stop(join_worker: true)
-          message_queue = @message_queue
-          message_queue << :close if message_queue
-
-          sender_thread = @sender_thread
-          sender_thread.join if sender_thread && join_worker
-        end
+      def stop()
+        return if @is_closed
+        # use this lock to both: not having another thread stopping this instance nor
+        # having a #add call creating a new thread
+        @mux.synchronize {
+          @is_closed = true
+          if sender_thread.alive? # no reasons to stop the bg thread is none is running already
+            blocking_queue = CloseQueue.new
+            channel << blocking_queue
+            blocking_queue.pop # wait for the bg thread to finish its work
+            blocking_queue.close if CLOSEABLE_QUEUES
+            sender_thread.join(3) # wait for completion, timeout after 3 seconds
+            # TODO(remy): should I close `channel` here?
```

Comment:
As I suggested above, I think it'd be simpler to just not use close; if we do, calling close here may be problematic if two stops get called concurrently, but the background thread is taking long to finish. E.g. something like T1: acquire mutex -> tell background thread to stop -> timeout join -> call close -> release mutex; T2: acquire mutex ->  does not see previous `@is_closed` -> tries to write to channel -> channel has been closed.

Reply:
Also, should the stop behavior be a bit more flexible? E.g. configurable timeout, or optionally run a block to decide what to do.